In [1]:
import math
import time
import hashlib
import itertools

# Computational Theory Tasks

## Task 1: Binary Representations

This task involves implementing four functions aimed to manipulate binary data, this is fundemental to many crythographic algorithms. Bitwise Operations allow for the manipulation of individual bits of data, these operations are applied directly to binary representations of numbers.  In cryptography bitwise operations are used to ensure confusion and diffusion in data.  Some common uses of these operations are in Hash functions, Encryption algorithms, Blockchain aswell as many others.[[1]](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf)

---------------------------------------------------------------------------

### 1. 'rotl(x, n=1)' 
This function rotates bits of a 32-bit unsigned integer 'x' to the left by 'n' places this is effectively shifting bits and wrapping the leftmost bits back into the rightmost position. The most common use for these functions are in crypographic hash functions such as SHA-1 and SHA-256 to provide bitwise diffusion making it difficult to reverse hash.  They are also used in applications like cylic buffers or rotating registers in hardware.[[3]](https://www.geeksforgeeks.org/rotate-bits-of-an-integer/)

In [2]:
def rotl(x, n=1):
    x &= 0xFFFFFFFF  #ensure x is a 32-bit unsigned integer
    return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF #rotate left by n bits

#### Step 1 Ensure x is a 32-bit unsigned integer.
- `0xFFFFFFFF` is the hexidecimal constant that represents a 32-bit mask.

- The bitwise AND (&) operation ensures that x only has the lower 32 bits, preventing unexpected large numbers.
#### Step 2 Rotate left by n bits
- `(x << n)` : Left shift moves bits n places to the left, filling with 0s, bits moved beyond 32nd bit are lost

- `(x >> (32 - n))` : Right shift moves the lost bits back to rightmost positions 

- `|` : Bitwise OR combines the shifted values.
### Step 3 Ensure Result is 32-bit
- `& 0xFFFFFFFF` : This ensures result stays within 32-bit unsigned integer range  
[[4]](https://realpython.com/python-bitwise-operators/) [[5]](https://www.linkedin.com/pulse/understanding-bitwise-operations-python-cryptography-hashing-woon-sm7ec/)

In [3]:
#example usage:
x = 0b01
n = 3
result = rotl(x, n)
print(f"Original: {x:032b}")
print(f"Rotated Left {n} positions: {result:032b}")

Original: 00000000000000000000000000000001
Rotated Left 3 positions: 00000000000000000000000000001000


---------------------------------------------------------------------------

### 2. 'rotr(x n=1)'
This function rotates bits of a 32-bit unsigned integer 'x' to the right 'n' places this is effectively shifting bits and wrapping the rightmost bits back into the leftmost position. This is similar to the 'rotl' function with the same use cases.[[3]](https://www.geeksforgeeks.org/rotate-bits-of-an-integer/)

In [4]:
def rotr(x, n=1):
    x &= 0xFFFFFFFF #ensure x is a 32-bit unsigned integer
    return ((x >> n) | (x << (32 - n))) & 0xFFFFFFFF #rotate right by n bits

#### Step 1 Ensure x is a 32-bit unsigned integer.
- `0xFFFFFFFF` is the hexidecimal constant that represents a 32-bit mask.

- The bitwise AND (&) operation ensures that x only has the lower 32 bits, preventing unexpected large numbers.
#### Step 2 Rotate right by n bits
- `(x >> n)` : Right shift moves bits n places to the right, filling with 0s, bits moved beyond 32nd bit are lost

- `(x << (32 - n))` : Left shift moves the lost bits back to leftmost positions 

- `|` : Bitwise OR combines the shifted values.
### Step 3 Ensure result is 32-bit
- `& 0xFFFFFFFF` : This ensures result stays within 32-bit unsigned integer range  
[[4]](https://realpython.com/python-bitwise-operators/) [[5]](https://www.linkedin.com/pulse/understanding-bitwise-operations-python-cryptography-hashing-woon-sm7ec/)

In [5]:
#example usage:
x = 0b01
n = 3
result = rotr(x, n)
print(f"Original: {x:032b}")
print(f"Rotated Right by {n} positions: {result:032b}")

Original: 00000000000000000000000000000001
Rotated Right by 3 positions: 00100000000000000000000000000000


---------------------------------------------------------------------------

### 3. 'def ch(x, y, z)'
This function chooses bits from two values: y and z, based off the corresponding bits in x. This works by choosing the corresponding bit in y if the x bit is 1; and choosing the bit corresponding bit in z if x is 0. This is a core compononent in the SHA-2(Secure Hash Algorithm) its used in the message schedule, it works with the bitwise operations to choose part of data from different sources based on values of x. Its also helps mix data from multiple sources to enhace unpredictability of cryptogrpahic transformations.[[1]](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf)

In [6]:
def ch(x, y, z):
     #choose y if x is true(1), otherwise choose z if x is false(0)
    return (x & y) ^ (~x & z)

#### Step 1 Select y where x is 1
- `(x & y)` : Bitwise AND(&) ensures that only the bits where x is 1 are taken from y
#### Step 2 Select z where x is 0
- `(~x & z) `: Bitwise NOT (~x) inverts bits in x, making the 1s to 0s and vice versa

- Bitwise AND(&) with z ensures only bits were x was originally 0 can be taken form z
#### Step 3 Combine results
- `(x & y) ^ (~x & z)` : Bitwise XOR(^) combines the two selections, ensuring bits from only one (y or z) is taken at each position   
[[4]](https://realpython.com/python-bitwise-operators/) [[5]](https://www.linkedin.com/pulse/understanding-bitwise-operations-python-cryptography-hashing-woon-sm7ec/)

In [7]:
#example usage:
x = 0b1100  
y = 0b1010  
z = 0b0110 
result = ch(x, y, z)
print(f"x: {x:04b}, y: {y:04b}, z: {z:04b}")
print(f"ch(x, y, z): {result:04b}")

x: 1100, y: 1010, z: 0110
ch(x, y, z): 1010


---------------------------------------------------------------------------

### 4. 'maj(x, y, z)'
This function takes a majority vote of the bits in each position in x, y and z. Output has 1 in position i if at least two of x, y, z have 1s in position i otherwise the result is 0. This is commonly used in Cryptographic hash functions such as SHA-256 and ensures output is influenced by majority of bits, increasing diffusion, improving resistance to attacks. Its also used in certain Error correction Codes or correction schemes where majority of values are trusted, allowing erros to be flagged based on outliers.[[1]](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf)

In [8]:
def maj(x, y, z):
    #choose the majority of x, y, z
    return (x & y) ^ (x & z) ^ (y & z)

#### Step 1 Calculate the majority between x and y
- `(x & y)` : Bitwise AND(&) between x and y give the bits where both x and y have 1 in the same position

- This operation shows a majority of the two bits at each position
#### Step 2 Calculate the majority between x and z
- `(x & z)` : Bitwise AND(&) between x and z gives the bits where both x and z have 1 in the same position
#### Step 3 Calculate the majority between y and z
- `(y & z)` : Bitwise AND(&) between y and z gives the bits where both x and z have 1 in the same position
#### Step 4 Combine results with XOR(^)
- `(x & y) ^ (x & z) ^ (y & z)` : Bitwise XOR(^) combines the results of these three majoritys. The XOR operation ensures only the majority value of the three will be choosen

- Eg. If two of the three bits are 1, the result will be 1 fot that bit position  
[[4]](https://realpython.com/python-bitwise-operators/) [[5]](https://www.linkedin.com/pulse/understanding-bitwise-operations-python-cryptography-hashing-woon-sm7ec/)


In [9]:
#example usage:
x = 0b1100  
y = 0b1010  
z = 0b0110 
result = maj(x, y, z)
print(f"x: {x:04b}, y: {y:04b}, z: {z:04b}")
print(f"maj(x, y, z): {result:04b}")

x: 1100, y: 1010, z: 0110
maj(x, y, z): 1110


---------------------------------------------------------------------------

### Test function
Testing each Binary operation function using hexdecimal number representations

In [10]:
#test cases
def test_binary_functions():
    #test rotl using hexadecimal numbers
    #0x01 = 0000 0001, 0x02 = 0000 0010
    assert rotl(0x01, 1) == 0x02
    #0x80000000 = 1000 0000 0000 0000 0000 0000 0000 0000, 0x00000001 = 0000 0000 0000 0000 0000 0000 0000 0001
    assert rotl(0x80000000, 1) == 0x00000001
    #0x12345678 = 0001 0010 0011 0100 0101 0110 0111 1000, 0x23456781 = 0010 0011 0100 0101 0110 0111 1000 0001 
    assert rotl(0x12345678, 4) == 0x23456781

    #test rotr using hexadecimal numbers
    #0x01 = 0000 0001, 0x02 = 0000 0010
    assert rotr(0x02, 1) == 0x01
    #0x80000000 = 1000 0000 0000 0000 0000 0000 0000 0000, 0x00000001 = 0000 0000 0000 0000 0000
    assert rotr(0x00000001, 1) == 0x80000000
    #0x12345678 = 0001 0010 0011 0100 0101 0110 0111 1000, 0x81234567 = 1000 0001 0010 0011 0100 0101 0110 0111
    assert rotr(0x12345678, 4) == 0x81234567

    #test ch using hexadecimal numbers
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0xFFFFFFFF, 0xAAAAAAAA, 0x55555555) == 0xAAAAAAAA #x is all ones, so the result is y
    #0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0x00000000, 0xAAAAAAAA, 0x55555555) == 0x55555555 #x is all zeros, so the result is z
    #0xF0F0F0F0 = 1111 0000 1111 0000 1111 0000 1111 0000, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0xF0F0F0F0, 0xAAAAAAAA, 0x55555555) == 0xA5A5A5A5 #0xA5A5A5A5 = 1010 0101 1010 0101 1010 0101 1010 0101 (combination of y and z as x is mixed)

    #test maj using hexadecimal numbers
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000
    assert maj(0xFFFFFFFF, 0x00000000, 0x00000000) == 0x00000000 # majority is 0x00000000 so the result is 0x00000000
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000
    assert maj(0xFFFFFFFF, 0xFFFFFFFF, 0x00000000) == 0xFFFFFFFF # majority is 0xFFFFFFFF so the result is 0xFFFFFFFF
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0xF0F0F0F0 = 1111 0000 1111 0000 1111 0000 1111 0000
    assert maj(0xFFFFFFFF, 0xF0F0F0F0, 0xF0F0F0F0) == 0xF0F0F0F0 # majority is 0xF0F0F0F0 so the result is 0xF0F0F0F0

    print("All test cases passed successfully")

test_binary_functions()

All test cases passed successfully


---------------------------------------------------------------------------

### Task 1 References:
1. FIPS PUB 180-4: Secure Hash Standard (SHS) :https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf  
2. M. A. Hossain, M. M. Alam, and M. S. Alam, "Cryptography Encryption Technique Using Circular Bit Rotation in Binary Field," 2020 IEEE Region 10 Symposium (TENSYMP), Dhaka, Bangladesh, 2020: https://ieeexplore.ieee.org/document/9197845/
3. Rotate bits of a number: https://www.geeksforgeeks.org/rotate-bits-of-an-integer/
4. RealPython: Bitwise operators in python: https://realpython.com/python-bitwise-operators/  
5. Understanding Bitwise Operations in Python: Cryptography, Hashing, and Real-World Applications: https://www.linkedin.com/pulse/understanding-bitwise-operations-python-cryptography-hashing-woon-sm7ec/

---------------------------------------------------------------------------

## Task 2: Hash Functions

### Polynomial Rolling Hash Function
Hashing is common practice in the world of computer science used to map data, such as strings, to numerical values. A widely used approach for hasing strings is the polynomial rolling hash function.  
This function is defined as:  
$$H(s) = (s_1 \times p^{(n-1)} + s_2 \times p^{(n-2)} + ... + s_n \times p^0) \mod m$$

- $H(s)$ is hash value of the string
- $s1, s2, sn$ are ASCII values of characters in the string
- $p$ is a prime base (commonly 31)
- $m$ is the modulus value (commonly 101)
- $n$ is the lenght of the string 

The Polynomial Rolling Hash Function has many use cases and is widely used in
- String Matching Algorithms
- Data Structures
- Plagiarism Detection
- Crypography  
[[1]](https://www.geeksforgeeks.org/string-hashing-using-polynomial-rolling-hash-function) [[2]](https://cp-algorithms.com/string/string-hashing.html)

In [11]:
def hash(s):
    #initialise the hash function to 0
    hashval = 0

    #iterate over each character in the string
    for c in s:
        #compute hash using polynomial rolling hash function
        hashval = ord(c) + 31 * hashval
        print(f"Character: {c}, Hash: {hashval}")
    
    #apply modulo operation to keep hash within fixed range (0-100)
    hashval =  hashval % 101

    return f"Final Hash Value: {hashval}"

### Code Explanation

- This variable will store the computed hash value, this will be happen as the function processes each charecter in the string.

In [12]:
hashval = 0

### Polynomial Rolling Hash Calculation
- This itereates over each charecter `c` in the string `s`. Each character in `s` contributes to the final hash.

- `ord(c)`: Converts the character `c` into ASCII value.

- `31 * hashval` : This takes the current hash and multiplies it by 31, this is a prime number commonly used for hash functions.

- `ord(c) + 31 * hashval` : The full line adds the ASCII of the character to a scaled hash value, this is to ensure each character incluences the hash in differently depending on the characters position. [[4]](https://medium.com/%40ggaappuu1234/understanding-rolling-hash-a-key-component-in-string-matching-algorithms-83236d8c4a20)

In [13]:
#string to be hashed
s = "hello"

#interate over each character in the hello
for c in s:
    #compute hash using polynomial rolling hash function
    hashval = ord(c) + 31 * hashval

print(f"Hash Value: {hashval}")

Hash Value: 99162322


### Modulo Operation (%101)
- This aims to ensure the final hash is in a specified range (0-100)

- This is to avoid faults like integer overflow, but also keeps the hash value compact.

- The prime modulous (101) is used as it helps distribute hash values evenly, which helps reduce collisions. [[6]](https://codeforces.com/blog/entry/60442)

In [14]:
hashval % 101

17

### Example of how function works:

In [15]:
print(hash("hello"))

Character: h, Hash: 104
Character: e, Hash: 3325
Character: l, Hash: 103183
Character: l, Hash: 3198781
Character: o, Hash: 99162322
Final Hash Value: 17


Iterate over each charecter in hello applying rolling hash calculation:
1. h (ASCII 104): hashval = 104
2. e (ASCII 101): hashval = 101 + 31 x 104 = 3325
3. l (ASCII 108): hashval = 108 + 31 x 3325 = 103183
4. l (ASCII 108): hashval = 108 + 31 x 103183 = 3198781
5. o (ASCII 111): hashval = 111 + 31 x 3198781 = 99162322

Apply modulo operation:
99162322 mod 101 = 17

"Hello" = 17

### Why use 31 as the Base?
1. Efficient Computation:
    - 31 is close to the powers of 2, this allows efficient multiplication and bitwise operations on many processors.
2. Empirical Evidence:
    - Many real-world applications use 31 due to it providing a good balance of speed and low collision probability.
3. Prime Number Property:
    - Prime numbers reduce chances of hash collisions occuring this is done by ensuring more uniform distribution of hash values [[6]](https://codeforces.com/blog/entry/60442)

### Why use 101 as the Modulus?
1. Prime Numbers Improve Hash Distribution:
    - Prime modulus ensures values are distributed evenly this inturn reduces hash collisions
2. Keeps Hash Values Compact:
    - Modulo operations keep hash values in a range in this case that range is 0 -100 this makes it more efficient for hashing tables.[[3]](https://www.designgurus.io/answers/detail/why-should-hash-functions-use-a-prime-number-modulus)

---------------------------------------------------------------------------

### Task 2 References:
1. String hasing using Polynomial rolling hash function: https://www.geeksforgeeks.org/string-hashing-using-polynomial-rolling-hash-function
2. String hashing: https://cp-algorithms.com/string/string-hashing.html
3. Why should hash functions use a prime number modulus?: https://www.designgurus.io/answers/detail/why-should-hash-functions-use-a-prime-number-modulus
4. Understanding Rolling Hash: A Key Component in String Matching Algorithms: https://medium.com/%40ggaappuu1234/understanding-rolling-hash-a-key-component-in-string-matching-algorithms-83236d8c4a20
5. Introduction to Rolling Hash – Data Structures and Algorithms: https://www.geeksforgeeks.org/introduction-to-rolling-hash-data-structures-and-algorithms/
6. On the mathematics behind rolling hashes and anti-hash tests: https://codeforces.com/blog/entry/60442

---------------------------------------------------------------------------

## Task 3: SHA256

SHA-256 is a cryptographic hash function that produces a 256 bit hash value, its part of the SHA-2 family designed by NSA and used as a federal standard(FIPS-180-4).[[1]](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf)

### Why Padding is Necessary
SHA-256 processes messages in 512 bit blocks. Messages must be padded to:
- Create complete 512 bit blocks
- Include the orignal message lenght
- Ensure the hash is computationally infeasible to reverse [[5]](https://medium.com/@bajrang1081siyag/how-sha-256-works-4951088ab9f8)

In [16]:
def sha256(file_path):
    #read the file contents
    with open(file_path, 'r') as f:
        binary_content = f.read().replace(' ', '')
    
    #convert binary string to bytes
    data = bytes(int(binary_content[i:i+8], 2) for i in range(0, len(binary_content), 8))
    
    #calculate original length in bits
    orig_len = len(data) * 8
    
    #start padding by adding the 1 bit (0x80 in hex)
    pad_data = data + b'\x80'
    
    #calculate padding length to make total length
    pad_len = (56 - (len(pad_data) % 64)) % 64
    
    #add zero padding
    pad_data += b'\x00' * pad_len
    
    #append original length as 64-bit big-endian unsigned integer
    pad_data += orig_len.to_bytes(8, 'big')
    
    #print padding in hex
    padding_hex = " ".join(f"{byte:02x}" for byte in pad_data[len(data):])
    print(padding_hex)
    

In [17]:
file_path = "test.txt"
sha256(file_path)

80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 18


### SHA-256 Padding Specification (FIPS-180-4)
The padding process consists of 3 steps:
#### Step 1: Append '1' bit
add a single '1' bit to the end of the message
- in byte format: 0x80 (100000000 in binary)
#### Step 2: Append '0' bits
Add enough 0 bits so the total lenght is congruent to 449 mod 512
- This leaves exactly 64 bits for the lenght information
- Formula: (448-(current_lenght + 1)) mod 512
#### Step 3: Append lenght
Add the orignal message lenght as a 64 bit big endian integer
- Represents the lenght in bits
- Allows messages up to 2^64 - 1 bits long

### Code explanation

In [18]:
# Step 1: Read and Process File
print("Step 1: Reading file contents")
#read the file contents
with open(file_path, 'r') as f:
    original_content = f.read()
    binary_content = original_content.replace(' ', '')

#convert to binary contents 
print (f"Original Content: {original_content}")
print (f"Binary Content: {binary_content}")

Step 1: Reading file contents
Original Content: 01100001 01100010 01100011
Binary Content: 011000010110001001100011


In [19]:
# Step 2: Convert Binary String to Bytes
print("Step 2: Converting binary string to bytes")
data = bytes(int(binary_content[i:i+8], 2) for i in range(0, len(binary_content), 8))

for i in range(0, len(binary_content), 8):
    binary_chunk = binary_content[i:i+8]
    byte_value = int(binary_chunk, 2)
    print(f"Binary: {binary_chunk} -> Decimal: {byte_value} -> Hex: {byte_value:02x}")

Step 2: Converting binary string to bytes
Binary: 01100001 -> Decimal: 97 -> Hex: 61
Binary: 01100010 -> Decimal: 98 -> Hex: 62
Binary: 01100011 -> Decimal: 99 -> Hex: 63


In [20]:
# Step 3: Calculate Original Length
print("Step 3: Calculating original length")
orig_len = len(data) * 8
print(f"Data length: {len(data)} bytes = {orig_len} bits\n")

Step 3: Calculating original length
Data length: 3 bytes = 24 bits



In [21]:
# Step 4: Add '1 Bit
print("Step 4: Adding '1' bit (0x80)")
pad_data = data + b'\x80'
print(f"0x80 in binary: {0x80:08b}")
print(f"Data after adding 0x80: {pad_data}\n")

Step 4: Adding '1' bit (0x80)
0x80 in binary: 10000000
Data after adding 0x80: b'abc\x80'



In [22]:
# Step 5: Calculate Zero Padding
print("Step 5: Calculating zero padding length")
pad_len = (56 - (len(pad_data) % 64)) % 64
print(f"Current length: {len(pad_data)} bytes")
print(f"Target length: 56 bytes (448 bits)")
print(f"Padding needed: {pad_len} bytes\n")

Step 5: Calculating zero padding length
Current length: 4 bytes
Target length: 56 bytes (448 bits)
Padding needed: 52 bytes



In [23]:
# Step 6: Add Zero Padding
print("Step 6: Adding zero padding")
pad_data += b'\x00' * pad_len
print(f"Data after zero padding: length = {len(pad_data)} bytes\n")

Step 6: Adding zero padding
Data after zero padding: length = 56 bytes



In [24]:
# Step 7: Add Length
print("Step 7: Adding original length as 64-bit big-endian")
length_bytes = orig_len.to_bytes(8, 'big')
print(f"Original length: {orig_len} bits")
print(f"As 64-bit big-endian: {' '.join(f'{b:02x}' for b in length_bytes)}")
pad_data += length_bytes
print(f"Final padded data length: {len(pad_data)} bytes\n")

Step 7: Adding original length as 64-bit big-endian
Original length: 24 bits
As 64-bit big-endian: 00 00 00 00 00 00 00 18
Final padded data length: 64 bytes



In [25]:
# Step 8: Print Final Padding
print("Step 8: Final padding in hex")
padding_hex = " ".join(f"{byte:02x}" for byte in pad_data[len(data):])
print(padding_hex)

Step 8: Final padding in hex
80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 18


### Example Analysis
For the message "abc" (binary: 01100001 01100010 01100011)
1. Orignal: 3 bytes = 24 bits
2. After '1' bit: 24 + 1 = 25 bits
3. Padding required: 488 - 25 = 423 bits
4. Zero padding: 423 bits = 52 bytes
5. Lenght to append: 8 bytes
6. Total: 3 + 1 + 52 + 8 = 64 bytes

Final padding is: 80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 18

---------------------------------------------------------------------------

### Task 3 References:
1. FIPS PUB 180-4: Secure Hash Standard (SHS): https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf  
2. Schneier, B.(2015). Applied Cryptography: https://www.schneier.com/books/applied-cryptography/
3. Menezes, A. et al.(1996) Handbook of Applied Cryptography: https://galois.azc.uam.mx/mate/propaganda/Menezes.pdf
4. NIST Special Publication 800-107: Recommendation for Applications Using Approved Hash Algorithms
5. How SHA-256 works?: https://medium.com/@bajrang1081siyag/how-sha-256-works-4951088ab9f8

---------------------------------------------------------------------------

## Task 4: Prime Numbers

### Algoritm 1: Sieve of Eratosthenes

### How it works
Find the first n prime numbers using the Sieve of Eratosthenes algorithm.

Algorithm Explanation
1. Create a boolean array for numbers from 2 to an estimated limit
2. Mark all multiples of each prime starting from 2
3. All unmarked numbers at the end are prime
4. Extract the first n prime numbers from the sieve

Advantages: 
- Very efficient for finding many primes in a range
- Significantly faster that trial division for large n
- Classic algorithm known for it efficiency

Disadvantages:
- Requires more memory than trial division
- Need to estimate upper limit for n primes  
[[1]](https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes) [[4]](https://www.baeldung.com/cs/prime-number-algorithms)

In [26]:
def sieve_of_eratosthenes(n):
    #estimate upper limit for n prime number
    #using prime number theorem approximation
    if n < 6:
        limit = 15
    else:
        limit = int(n * (math.log(n) + math.log(math.log(n))))

    #create sieve array - true means prime
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False #0 and 1 are not prime

    #mark multiples of each prime
    for i in range(2, int(math.sqrt(limit)) + 1):
        if sieve[i]:
            #mark all multiples of i as not prime
            for j in range(i * i, limit + 1, i):
                sieve[j] = False

    #collect prime numbers
    primes = []
    current = 2
    while len(primes) < n and current <= limit:
        if sieve[current]:
            primes.append(current)
        current += 1
    return primes

In [27]:
print("Finding first 100 prime numbers using Sieve of Eratosthenes")
start_time = time.time()
n = 100
primes = sieve_of_eratosthenes(n)
print(f"First {n} prime numbers: {primes} \nin {time.time() - start_time:.4f} seconds")

Finding first 100 prime numbers using Sieve of Eratosthenes
First 100 prime numbers: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541] 
in 0.0002 seconds


---------------------------------------------------------------------------

### Algorithm 2: Trial Division Method

### How it works
Find the first n prime numbers using trial division algorithm

Algorithm Explanation:
1. Start checking from 2 (first prime)
2. For each number check if its divisible by any number from 2 to sqrt(n)
3. If no divisors found its a prime
4. Continue until we find n primes

Advantages:
- Simple to implement
- Small memory requirment
- Can find primes one by one as needed

Disadvantages:
- Slower that other algorithms for large n
- Redundant calculations (checks same number multiple times) [[3]](https://www.khanacademy.org/computing/computer-science/cryptography/comp-number-theory/a/trial-division)

In [28]:
def trial_division(n):
    primes = []
    current = 2
    while len(primes) < n:
        is_prime = True
        #optimise by only checking to square root of current num
        for i in range(2, int(math.sqrt(current))+1):
            if current % i == 0:
                is_prime = False
                break
        if is_prime:
            primes.append(current)
        #optimise after 2 only check odd numbers
        current += 1 if current == 2 else 2

    return primes

In [29]:
print("Finding first 100 prime numbers using trial division")
start_time = time.time()
n = 100
primes = trial_division(n)
print(f"First {n} prime numbers: {primes} \nin {time.time() - start_time:.4f} seconds")

Finding first 100 prime numbers using trial division
First 100 prime numbers: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541] 
in 0.0005 seconds


---------------------------------------------------------------------------

### Task 4 References

1. Sieve of Eratosthenes: https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes
2. Prime Number in Python: https://www.ccbp.in/blog/articles/prime-number-in-python
3. Trial division(article) Khan academy : https://www.khanacademy.org/computing/computer-science/cryptography/comp-number-theory/a/trial-division
4. Fastest Algoritm to Find Prime Numbers, Baeldung: https://www.baeldung.com/cs/prime-number-algorithms
5. Check for Prime Number, GeeksforGeeks: https://www.geeksforgeeks.org/check-for-prime-number/


---------------------------------------------------------------------------

## Task 5: Roots

This task requires the calculation of the first 32 bits of the fractional part of the square root of the first 100 prime numbers. When we take the square root of a number we get a irrational result whit an integer part and a fractional part. For example sqrt(2) = 1.41421.... where 1 is the interger part and 0.41421 is the fractional part. In this task we need to extract the fractional part representing the first 32 bits in binary format. [[1]](https://stackoverflow.com/questions/4674956/what-are-the-first-32-bits-of-the-fractional-part-of-this-float)

In [30]:
def fractional_bits_of_sqrt(n, bits=32):
    #calculate square root
    sqrt = math.sqrt(n)
    
    #extract fractional part
    integer = int(sqrt)
    fraction = sqrt - integer

    #convert to binary
    binary_bits = []
    current_fraction = fraction
    for _ in range(bits):
        #multiply by 2
        current_fraction *= 2
        #get integer part
        bit = int(current_fraction)
        binary_bits.append(str(bit))
        #keep only the fractional part
        current_fraction -= bit

    return ''.join(binary_bits)
        

In [31]:
def get_bits(primes):
    results = []
    #iterate over each prime number
    for prime in primes:
        #get 32 bits of fractional part
        binary_fraction = fractional_bits_of_sqrt(prime, 32)

        results.append({
            'prime': prime,
            'sqrt_fraction': binary_fraction,
        })
    return results

In [32]:
primes = sieve_of_eratosthenes(100)
sqrt_bits = get_bits(primes)
print("Fractional bits of square root of first 100 prime numbers:")
for result in sqrt_bits:
    prime = result['prime']
    sqrt_fraction = result['sqrt_fraction']
    print(f"Prime: {prime}, Fractional bits of sqrt: {sqrt_fraction}")

Fractional bits of square root of first 100 prime numbers:
Prime: 2, Fractional bits of sqrt: 01101010000010011110011001100111
Prime: 3, Fractional bits of sqrt: 10111011011001111010111010000101
Prime: 5, Fractional bits of sqrt: 00111100011011101111001101110010
Prime: 7, Fractional bits of sqrt: 10100101010011111111010100111010
Prime: 11, Fractional bits of sqrt: 01010001000011100101001001111111
Prime: 13, Fractional bits of sqrt: 10011011000001010110100010001100
Prime: 17, Fractional bits of sqrt: 00011111100000111101100110101011
Prime: 19, Fractional bits of sqrt: 01011011111000001100110100011001
Prime: 23, Fractional bits of sqrt: 11001011101110111001110101011101
Prime: 29, Fractional bits of sqrt: 01100010100110100010100100101010
Prime: 31, Fractional bits of sqrt: 10010001010110010000000101011010
Prime: 37, Fractional bits of sqrt: 00010101001011111110110011011000
Prime: 41, Fractional bits of sqrt: 01100111001100110010011001100111
Prime: 43, Fractional bits of sqrt: 100011101011

### Code Explanation

#### Step 1: Getting the square root.

In [33]:
sqrt = math.sqrt(2)
print(f"Square root of 2: {sqrt}")

Square root of 2: 1.4142135623730951


#### Step 2: Extracting fractional part

In [34]:
integer_part = int(sqrt)
print(f"Integer part: {integer_part}")
fractional_part = sqrt - integer_part
print(f"Fractional part: {fractional_part}")

Integer part: 1
Fractional part: 0.41421356237309515


We separate the integer and fractional parts. ef
- integer part = 1
- frational part = 0.41421356237309515  
  
as shown above in the code example

#### Step 3: Converting fractional part to Binary

In [35]:
binary_bits = []
current_fraction = fractional_part

for _ in range(32):
    #multiply by 2
    current_fraction *= 2
    #get integer part
    bit = int(current_fraction)
    binary_bits.append(str(bit))
    #keep only the fractional part
    current_fraction -= bit

print(f"Binary bits: {''.join(binary_bits)}")


Binary bits: 01101010000010011110011001100111


This is the core of the algorithm for converting decimal fraction to binary:
1. Multiply fraction by 2
2. Take the interger part as the next binary digit
3. Keep the remaining fractional part
4. Repeat for all 32 bits  
  
For example with 0.414....
- 0.414 x 2 = 0.0828 = bit is 0
- 0.828 x 2 = 1.656 = bit is 1
- 0.656 x 2 = 1.312 = bit is 1
- etc.


We join all the binary digits into a string and return it.

---------------------------------------------------------------------------

### Task 5 References
1. What are the first 32 bits of the fractional part of this float?, Stack Overflow: https://stackoverflow.com/questions/4674956/what-are-the-first-32-bits-of-the-fractional-part-of-this-float
2. Fractional parts of square roots of primes, Number theory: https://math.stackexchange.com/questions/3185456/fractional-parts-of-square-roots-of-primes
3. Extract Specific Bits of an Integer, Spiceworks:  https://community.spiceworks.com/t/extract-specific-bits-of-an-integer/845522

---------------------------------------------------------------------------

## Task 6: Proof of Work

This task requires us to find the words in English with the greatest number of leading zero bits at the beginning of their SHA-256 hash digest. This is a important part of the cryptographic mining problem, which is similar to what cryptocurrencies like Bitcoin use in their proof of work systems.  
  
A SHA-256 hash is a 256 bit value. Each input (word) produces a hash output. The objective is to find the word the words that produces hashs with many leading zeros when the hash is represented in binary.  
[[1]]( https://www.investopedia.com/terms/p/proof-work.asp) [[2]](https://www.britannica.com/money/proof-of-work-blockchain-verification)

Orignally I used a function to download the following dictionaries:
- https://raw.githubusercontent.com/dwyl/english-words/master/words_alpha.txt
- https://www.wordgamedictionary.com/sowpods/download/sowpods.txt  
  
However these dictionaries proved to be inaccurate when words were compared with both the oxford and cambridge dictionaries this is why I chose to use the words.txt file from last semester as after alot of searching its the only file I can find that has accurate words.  
While this dictionary is missing some words it covers a large number of the words within the dictionaries outlined above. I added the word goaltenders as from previous tests this showed to be the highest.

In [36]:
def count_leading_zeros(word):
    #get the SHA-256 hash of the word
    hash_object = hashlib.sha256(word.encode('utf-8'))
    hash_digest = hash_object.digest()

    #count leading zeros
    leading_zeros = 0
    for byte in hash_digest:
        #process each byte
        for bit_position in range(8):
            #check if the bit is 0
            if not (byte & (128 >> bit_position)):
                leading_zeros += 1
            else:
                #stop counting at first 1
                return leading_zeros
    #return total leading zeros
    return leading_zeros

In [37]:
def most_leading_zeros(n=1):
    #set the dictionary file path
    dictionary_file = 'words.txt'
    #open dictionary file
    with open(dictionary_file, 'r') as f:
        words = [line.strip() for line in f if line.strip()]

    word_zeros = []

    #iterate over each word
    for word in words:
        #make all words lowercase has more leading zeros
        word = word.lower()
        #count leading zeros
        zeros = count_leading_zeros(word)
        word_zeros.append((word, zeros))

    #sort by leading zeros
    word_zeros.sort(key=lambda x: x[1], reverse=True)

    return word_zeros[:n]

In [38]:
top_10_words = most_leading_zeros(5)
print("Top 5 words with most leading zeros:")
for word, zeros in top_10_words:
    print(f"Word: {word}, Leading Zeros: {zeros}")

Top 5 words with most leading zeros:
Word: goaltenders, Leading Zeros: 18
Word: grady, Leading Zeros: 15
Word: mountable, Leading Zeros: 15
Word: courteously, Leading Zeros: 14
Word: teared, Leading Zeros: 14


### Proof words are in dictionary
- goaltenders: https://www.oed.com/dictionary/goaltender_n?tab=factsheet#2790795
- grady: https://www.oed.com/dictionary/grady_adj?tab=factsheet#2638712
- mountable: https://www.oed.com/dictionary/mountable_adj?tab=factsheet#35743521
- courteously: https://www.oed.com/dictionary/courteously_adv?tab=factsheet#8135742
- teared: https://www.oed.com/dictionary/tear_v1?tab=factsheet#19076903`

### Code Explanation

#### Step 1: Converting a word to SHA-256 Hash

In [39]:
hash_object = hashlib.sha256(b'hello')
hash_digest = hash_object.digest()

print(f"SHA-256 hash of 'hello': {hash_digest.hex()}")

SHA-256 hash of 'hello': 2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824


- word.encode('utf-8') : This converts the string to bytes
- hashlib.sha256(...) : This creates a hash object using the SHA-256 algorithm
- .digest() : Returns the hash as raw bytes insted of hexidecimal

###

#### Step 2: Counting leading zero bits in the hash

In [40]:
def example():
    leading_zeros = 0
    for byte in hash_digest:
        print(f"Byte: {byte:08b}")
        for bit_position in range(8):
            print(f"Bit position: {bit_position}, Bit value: {(byte & (128 >> bit_position)) >> (7 - bit_position)}")
            # Check if the current bit is 0
            if not (byte & (128 >> bit_position)):
                leading_zeros += 1
            else:
                # Stop at the first 1 bit
                return leading_zeros
    
print(f"Leading zeros in SHA-256 hash: {example()}")

Byte: 00101100
Bit position: 0, Bit value: 0
Bit position: 1, Bit value: 0
Bit position: 2, Bit value: 1
Leading zeros in SHA-256 hash: 2


Here we examine each byte in the hash digest and within each byte we must examine each bit from left to right. 
- 128 is binary 10000000
- 128 >> bit_position : shifts this bit to check each position
- byte & (128 >> bit_postion) : This tests if that bit is set to 1
- If bit is 0 we increment our counter
- As soon as we find a 1 bit, we return the count

#### Step 3: Processing each word

In [41]:
words = ["hello", "world", "it", "is", "a", "beautiful", "day"]
word_zeros = []
for word in words:
    #make word lowercase for consistency
    word = word.lower()
    #count leading zeros
    zeros = count_leading_zeros(word)
    #store word and its leading zeros
    word_zeros.append((word, zeros))

print("Words and their leading zeros:")
for word, zeros in word_zeros:
    print(f"Word: {word}, Leading Zeros: {zeros}")


Words and their leading zeros:
Word: hello, Leading Zeros: 2
Word: world, Leading Zeros: 1
Word: it, Leading Zeros: 2
Word: is, Leading Zeros: 0
Word: a, Leading Zeros: 0
Word: beautiful, Leading Zeros: 0
Word: day, Leading Zeros: 0


#### Step 4: Finding words with most zeros

In [42]:
word_zeros.sort(key=lambda x: x[1], reverse=True)
print("Sorted words by leading zeros:")
for word, zeros in word_zeros:
    print(f"Word: {word}, Leading Zeros: {zeros}")

Sorted words by leading zeros:
Word: hello, Leading Zeros: 2
Word: it, Leading Zeros: 2
Word: world, Leading Zeros: 1
Word: is, Leading Zeros: 0
Word: a, Leading Zeros: 0
Word: beautiful, Leading Zeros: 0
Word: day, Leading Zeros: 0


- key=lambda x: x[1] : This sorts based on the second element
- reverse=True : Sorts in decending order

---------------------------------------------------------------------------

### Task 6 References
1. What Is Proof of Work (PoW) in Blockchain?, Investopedia:  https://www.investopedia.com/terms/p/proof-work.asp
2. What is proof of work? Explaining blockchain verification, Britanica Money: https://www.britannica.com/money/proof-of-work-blockchain-verification
3. Are the leading zero's at the beginning of a hash sought after, if so why?, Reddit: https://www.reddit.com/r/cryptography/comments/1emup7k/are_the_leading_zeros_at_the_beginning_of_a_hash/
4. SHA-512 - How difficult is it to find a hash digest beginning with at least twelve zeros?, Crythography: https://crypto.stackexchange.com/questions/89690/sha-512-how-difficult-is-it-to-find-a-hash-digest-beginning-with-at-least-twel
5. Extracting a particular range of bits and find number of zeros between them in C, Stackoverflow: https://stackoverflow.com/questions/48659216/extracting-a-particular-range-of-bits-and-find-number-of-zeros-between-them-in-c

---------------------------------------------------------------------------

## Task 7: Turing Machines

A turing machine is a mathematical model of computation. It was developed by Alan Turing in 1936. It consists of a tape divided into cells, a read/write head and a set of rules which determine its behaviour based on the state and symbol.[[1]](https://www.cs.virginia.edu/~robins/Turing_Paper_1936.)

### Components
#### States(Q) 
The set {'R', 'A', 'C', 'E', 'H'} defines the machines possible states each has specific purpose
- R: Scan Right - move right until the end of the number
- A: Add 1 - add 1 to current bit
- C: Carry - handle carry
- E: Extend - add new digit
- H: Halt - machine is done

#### Tape Alphabet(T)
{'0', '1', '_'} represents all possible symbols on the tape including the binary digits and blank symbols

#### Blank symbol(b)
'_' shows empty cells on the tape

#### Input Alphabet(A)
{'0', '1'} contains the symbols that can appear in the input

#### Tranistion Function (state_tabel)
This defines the machines behaviour by mapping the current state/symbol to the next state/symbol to write.

#### Inital State(s) 
'R' is where computation begins

#### Final States(F)
{'H', 'E'} indicates when the machine should stop  
[[5]](https://stackoverflow.com/questions/2099650/designing-a-turing-machines-state-table) [[6]](https://www.geeksforgeeks.org/construct-turing-machine-for-incrementing-binary-number-by-1/)



In [43]:
#states
Q = {'R', 'A', 'C', 'E', 'H'}

#tape alphabet
T = {'0', '1', '_'}

#blank symbol
b = '_'

#input alphabet
A = {'0', '1'}

state_table = {
    #state R: Scan right until the end of the number
    ('R', '0'): ('R', '0', 'R'),  #keep moving right
    ('R', '1'): ('R', '1', 'R'),  #keep moving right
    ('R', '_'): ('A', '_', 'L'),  #end reached move back to last digit
    
    #state A: Add 1 to the current position
    ('A', '0'): ('H', '1', 'N'),  #0+1=1, we're done
    ('A', '1'): ('C', '0', 'L'),  #1+1=0 with carry
    ('A', '_'): ('E', '1', 'N'),  #empty tape case or need to extend
    
    #state C: Handle carry
    ('C', '0'): ('H', '1', 'N'),  #0+1=1, we're done
    ('C', '1'): ('C', '0', 'L'),  #1+1=0 with carry
    ('C', '_'): ('E', '1', 'N'),  #need to extend with a new digit
}

#initial state
s = 'R'
#final states
F = {'H', 'E'}

#### Binary Addition with Turing Machines
Adding 1 to a binary number follows the following rules:
- If the right most bit is 0, change it to 1 and finish
- If the right most bit is 1, change it to 0 and carry 1 to the next position
- Continue until theres no carry or new digit [[3]](https://medium.com/@vardanagarwal16/implementing-a-turing-machine-easily-in-python-55213fc8d5d5)

In [44]:
def turing_machine(table, input, verbose = True):
    if verbose is True:
        print(f"Input: {input}")

    #keep track of operations
    ops = 0

    #check for empty input
    if input == '':
        input = '_'

    #make input a list
    tape = list(input)

    #track position on tape
    pos = 0

    #initial state
    state = s
    
    if verbose is True:
        #print config
        print (state + ''.join(tape))

    
    #iterate until at final state
    while state not in F:
        #get current symbol
        symbol = tape[pos]

        #get next state symbol and direction
        if (state, symbol) in table:
            next_state, next_symbol, direction = table[(state, symbol)]

            #update tape
            tape[pos] = next_symbol

            #increment operations
            ops += 1

            #move in direction
            if direction == 'L':
                pos -= 1
            elif direction == 'R':
                pos += 1
            #n means no movement
            
            #fix finite list promblem
            if pos < 0:
                tape = [b] + tape
                pos = 0
            if pos == len(tape):
                tape.append('_')

            #update state
            state = next_state

            if verbose is True:
                #print config
                print (''.join(tape[:pos]) + state + ''.join(tape[pos:]))

        else:
            if verbose is True:
                print(f"Invalid state: {state}, symbol: {symbol}")
            break

    if verbose is True:
        print(f"Number of operations: {ops}")
        print(f"Output: {''.join(tape).strip("_")}")


In [45]:
example_input = '100111'
turing_machine(state_table, example_input)

Input: 100111
R100111
1R00111
10R0111
100R111
1001R11
10011R1
100111R_
10011A1_
1001C10_
100C100_
10C0000_
10H1000_
Number of operations: 11
Output: 101000


### Code explanation

#### Step 1: Setup

In [46]:
verbose = True
input = '100111'

if verbose is True:
    print(f"Input: {input}")
#keep track of operations
ops = 0
#check for empty input
if input == '':
    input = '_'
#make input a list
tape = list(input)
#track position on tape
pos = 0
#initial state
state = s

Input: 100111


This first part:
- Sets up the initial conditions
- Prepares the tape converting input string to a list for cell manipulation
- Initialises the head position to the leftmost cell
- Sets the starting state to the initial state 's'

#### Step 2: Main execution loop

In [47]:
table = state_table 

if verbose is True:
    #print config
    print (state + ''.join(tape))
#iterate until at final state
while state not in F:
    #get current symbol
    symbol = tape[pos]
    #get next state symbol and direction
    if (state, symbol) in table:
        next_state, next_symbol, direction = table[(state, symbol)]
    break #break just for example


R100111


This part:
- Prints the inital configuration if verbose is on
- Begins the main exectuion loop that continues until the state matches the final state
- Reads the current symbol from the tape at head position
- Looks up what to do next in the table based on state and symbol

#### Step 4: Update tape and position

In [48]:
#update tape
tape[pos] = next_symbol
#increment operations
ops += 1
#move in direction
if direction == 'L':
    pos -= 1
elif direction == 'R':
    pos += 1
#n means no movement

Here the machine:
- Writes the new symbol to the current position
- Counts this operation
- Moves the head left or right based on the direction from table or stays if 'N'

#### Step 5: Handle infinite tape

In [49]:
#fix finite list promblem
if pos < 0:
    tape = [b] + tape
    pos = 0
if pos == len(tape):
    tape.append('_')

This part is critical:
- Handles tapes infinite nature by extending it when needed
- If the head moves left of the tapes start a blank symbol is added at the start
- If the head moves right beyond the tapes end a blank symbol is added to the end

#### Step 6: Upadate state and display

In [50]:
#update state
state = next_state
if verbose is True:
    #print config
    print(''.join(tape[:pos]) + state + ''.join(tape[pos:]))

1R00111


Here:
- Update the current state to the next state
- Print the new config if in verbose mode showing the state at the head position

#### Step 7: Error handling and result output

In [51]:
if verbose is True:
    state = state # just so example works
else:
    if verbose is True:
        print(f"Invalid state: {state}, symbol: {symbol}")

if verbose is True:
    print(f"Number of operations: {ops}")
    print(f"Output: {''.join(tape).strip('_')}")

Number of operations: 1
Output: 100111


The final part:
- Handles error cases whre no transition is defined
- Prints operation count and final output

---------------------------------------------------------------------------

### Task 7 References:
1. Turing, A.M(1936) "On Computable Numbers": https://www.cs.virginia.edu/~robins/Turing_Paper_1936.
2. Sipser, M.(2013) "Introduction to the Theory of Computation" :https://drive.uqu.edu.sa/_/mskhayat/files/MySubjects/20189FS%20ComputationTheory/Introduction%20to%20the%20theory%20of%20computation_third%20edition%20-%20Michael%20Sipser.pdf
3. Implementing a Turing machine easily in Python, Medium: https://medium.com/@vardanagarwal16/implementing-a-turing-machine-easily-in-python-55213fc8d5d5
4. Turing Machine Simulator Using Python, Geeksforgeeks: https://www.geeksforgeeks.org/turing-machine-simulator-using-python/
5. Designing a Turing Machine's state table, Stackoverflow: https://stackoverflow.com/questions/2099650/designing-a-turing-machines-state-table
6. Construct Turing Machine for incrementing Binary Number by 1, Geeksforgeeks: https://www.geeksforgeeks.org/construct-turing-machine-for-incrementing-binary-number-by-1/

---------------------------------------------------------------------------

## Task 8: Computational Complexity

### Bubble sort
Bubble sort is a sorting algorithm. Named for the way smaller elements "bubble" to the top of the list. 

#### Time Complexity
- best case: O(n) (when array is already sorted)
- average case: O(n^2)
- worst case: O(n^2)

#### Space Complexity
- O(1), sorts in place

#### Advantages:
- Simple to implement
- Requires minimal memory
- Performs well on small datasets or nearly sorted arrays
- Can detect a sorted array in a single pass

#### Disadvantages:
- Inefficient for large datasets
- Performs more comparisons and sways then nessessary
- Outperformed by other algorithms like Quick sort  
[[1]](https://www.geeksforgeeks.org/bubble-sort-algorithm/) [[2]](https://stackoverflow.com/questions/104420/how-do-i-generate-all-permutations-of-a-list)

In [52]:
def bubble_sort(array):
    #create a of array to avoid modifying original
    array = array.copy()
    n = len(array)
    comparisons = 0

    #iterate over each element
    for i in range(n):
        #flag to check if any elements were swapped
        swapped = False
        #last i elements are already sorted
        for j in range(0, n-i-1):
            comparisons += 1
            #swap if element is greater than next
            if array[j] > array[j+1]:
                array[j], array[j+1] = array[j+1], array[j]
                swapped = True
        #if no elements were swapped break
        if not swapped:
            break

    return array, comparisons 

In [53]:
L = [1,2,3,4,5]
#generate all permutations of list
all_permutations = list(itertools.permutations(L))

#sort each permutation
for perm in all_permutations:
    #convert tuple to list
    perm_list = list(perm)
    #keep original permutation for printing
    original = perm_list.copy()

    _,comparisons = bubble_sort(perm_list)
    print(f"Permutation: {original}, Comparisons: {comparisons}")

Permutation: [1, 2, 3, 4, 5], Comparisons: 4
Permutation: [1, 2, 3, 5, 4], Comparisons: 7
Permutation: [1, 2, 4, 3, 5], Comparisons: 7
Permutation: [1, 2, 4, 5, 3], Comparisons: 9
Permutation: [1, 2, 5, 3, 4], Comparisons: 7
Permutation: [1, 2, 5, 4, 3], Comparisons: 9
Permutation: [1, 3, 2, 4, 5], Comparisons: 7
Permutation: [1, 3, 2, 5, 4], Comparisons: 7
Permutation: [1, 3, 4, 2, 5], Comparisons: 9
Permutation: [1, 3, 4, 5, 2], Comparisons: 10
Permutation: [1, 3, 5, 2, 4], Comparisons: 9
Permutation: [1, 3, 5, 4, 2], Comparisons: 10
Permutation: [1, 4, 2, 3, 5], Comparisons: 7
Permutation: [1, 4, 2, 5, 3], Comparisons: 9
Permutation: [1, 4, 3, 2, 5], Comparisons: 9
Permutation: [1, 4, 3, 5, 2], Comparisons: 10
Permutation: [1, 4, 5, 2, 3], Comparisons: 9
Permutation: [1, 4, 5, 3, 2], Comparisons: 10
Permutation: [1, 5, 2, 3, 4], Comparisons: 7
Permutation: [1, 5, 2, 4, 3], Comparisons: 9
Permutation: [1, 5, 3, 2, 4], Comparisons: 9
Permutation: [1, 5, 3, 4, 2], Comparisons: 10
Permu

### Code Explanation

In [54]:
array = [64, 34, 25, 12, 22, 11, 90]
#create a copy of array to avoid modifying original
array = array.copy()
#get length of array
n = len(array)
#initialise comparisons
comparisons = 0

This section sets up the function and initialises variables.
- We get a copy of the array so not to modify the original
- We get the lenght of the array to define the passes through the array
- We initialise the comaparisons variable which will store the number of comparisons per permutation

In [55]:
#iterate over each element
for i in range(n):
    #flag to check if any elements were swapped
    swapped = False

This outer loop makes n passes through the array if it does not hit a break. The swapped flag helps optimise by exiting early if the array becomes sorted

In [56]:
 #last i elements are already sorted
for j in range(0, n-i-1):
    comparisons += 1

The inner loop compares adjacent elements. We stop at n-i-1 because after each pass the largest i elements are already in their correct positions at the end of the array

In [57]:
j=0
#swap if element is greater than next
if array[j] > array[j+1]:
    array[j], array[j+1] = array[j+1], array[j]
    swapped = True

Here the actual comparison and swap takes place if an element is greater then the one to its right, we swap them and set our flag to indicate a swap occured.

In [58]:
#for loop as example
for example in array:
    #if no elements were swapped break
    if not swapped:
        break

This is the optimisation if no swaps occur during a pass then the array is sorted and we can exit early.

---------------------------------------------------------------------------

### Task 8 references
1. Bubble Sort Algorithm, GeeksforGeeks: https://www.geeksforgeeks.org/bubble-sort-algorithm/
2. How do I generate all permutations of a list?, Stackoverflow: https://stackoverflow.com/questions/104420/how-do-i-generate-all-permutations-of-a-list
3. 6.7. The Bubble Sort, pythonds: https://runestone.academy/ns/books/published/pythonds/SortSearch/TheBubbleSort.html

---------------------------------------------------------------------------

## End